### PR TITLE
🔒 Fixes for multiple JavaScript vulnerabilities

### PR DESCRIPTION
**Pull request by AlphaSecQ**
**Describe Pull request**
This pull request fixes multiple issues found inside `qb-atms.js` (hereafter referred to as "script"). These issues/vulnerabilities can be exploited through FiveM Developer Tools (Chrome Inspector) by directly modifying the JavaScript files.

### JavaScript Comparison Bypass
The first vulnerability is caused by the comparison of PIN number inside the script. Since all the JavaScript files are loaded client-side, we should avoid doing any important checks or comparisons in them without double checking it with server. Since we currently do PIN authentication purely inside the script, we can easily modify the line below to **ALWAYS** give us access into any card we use.

```js
if (currentVal == clientPin)  // Current comparison (vulnerable)

if (true)  // Exploited line (always grants access)
if (clientPin == clientPin)  // Exploited line (always grants access)
```

To fix this issue, I removed all the comparisons and authentication attempts inside the script and moved them to server-side callback named `qb-atms:server:verifyPIN` that can be found at the bottom of `server/main.lua`

### Custom post request to NUI Callback
After (or even before) fixing the issue above, we are left with another vulnerability. Since the NUI callback named `loadBankingAccount` inside `client/main.lua` doesn't verify whether the user is authorized to use the card, we can craft a custom post request that will grant us access to **ANY** credit card that is registered in the server. All we need is Citizen ID of the owner and the Card Number.

After we get that, we can craft a request like this and gain immediate access to the card
```js
$.post("https://qb-atms/loadBankingAccount", JSON.stringify({
    cid: "DKH47814",
    cardnumber: "6122691686594159"
}));
```

To fix this issue, the `server/main.lua` now has a table called `authorized` that stores which players are authorized to use certain cards. If the user is not in there and tries to load the card anyway (custom post request) it now doesn't let them through. This is all thanks to this line.

```lua
if not authorized[source] or authorized[source] ~= data.cardnumber then return end
```

<br>

## Infinite money
There is another method that could potentially allow two users to get unlimited money on their bank account.

Due to severity of this issue, I will not disclose any further information on how it works in order to prevent exploitation of this vulnerability, but I would be happy to share it either in private chat (with staff only of course) or once this request is merged (since it fixes this issue) 

<br>

---

**Questions (please complete the following information):**
- Have you personally loaded this code into an updated qbcore project and checked all it's functionality? **yes**
- Does your code fit the style guidelines? **yes**
- Does your PR fit the contribution guidelines? **yes**

### CODE REVIEW

**Comment by Tiimon** [comment]
Tested with newest qb-atms commits and it doesn't work. When a player enters their correct pin nothing happens

**Comment by GhzGarage** [comment]
Has conflicts now and an error reported above, please fix

File: server/main.lua [hunk 1]
```diff
@@ -122,6 +128,8 @@ RegisterNetEvent('qb-atms:server:doAccountWithdraw', function(data)
     end
 end)
 
+RegisterNetEvent('qb-atms:server:unathorize', function() authorized[source] = nil end)
```
Comment (ItsANoBrainer) [inline server/main.lua:131]:
```suggestion
RegisterNetEvent('qb-atms:server:unauthorize', function() authorized[source] = nil end)
```

File: client/main.lua [hunk 2]
```diff
@@ -76,6 +76,7 @@ end)
 -- Callbacks
 
 RegisterNUICallback("NUIFocusOff", function(data, cb)
+    TriggerServerEvent('qb-atms:server:unathorize')
```
Comment (ItsANoBrainer) [inline client/main.lua:79]:
```suggestion
    TriggerServerEvent('qb-atms:server:unauthorize')
```

**Comment by ADRNALN** [comment]
Tested, this is now working properly with the newest qb-atms update

**Comment by AlphaSecQ** [comment]
Merge issues are now resolved.
The bug should also be resolved (based on a latest comment from @ADRNALN)

File: server/main.lua [hunk 3]
```diff
@@ -227,3 +237,22 @@ QBCore.Functions.CreateCallback('qb-atms:server:loadBankAccount', function(sourc
...
+
+QBCore.Functions.CreateCallback('qb-atms:server:verifyPIN', function(source, cb, data)
+    local src = source
+    local xPlayer = QBCore.Functions.GetPlayer(src)
+    local visas = xPlayer.Functions.GetItemsByName('visa')
+    local masters = xPlayer.Functions.GetItemsByName('mastercard')
+    local returnValue = false
+
+    for i,v in pairs(visas) do
+        returnValue = returnValue and true or (v.info.cardNumber == data.cardNumber and v.info.cardPin == data.cardPin)
```
Comment (BerkieBb) [inline server/main.lua:249]: The `returnValue and true or` here is unnecessary, you can just make that `returnValue or`

File: server/main.lua [hunk 4]
```diff
@@ -227,3 +237,22 @@ QBCore.Functions.CreateCallback('qb-atms:server:loadBankAccount', function(sourc
     end
     cb(banking)
 end)
+
+QBCore.Functions.CreateCallback('qb-atms:server:verifyPIN', function(source, cb, data)
+    local src = source
+    local xPlayer = QBCore.Functions.GetPlayer(src)
+    local visas = xPlayer.Functions.GetItemsByName('visa')
+    local masters = xPlayer.Functions.GetItemsByName('mastercard')
```
Comment (BerkieBb) [inline server/main.lua:245]: These items check should have a check for empty tables, either by using `next(table)` or by using `table.type(table) == 'empty'` (table.type is a cfxlua function)

**Comment by Z3rio** [comment]
PR abandoned by creator.
Just reopen if you want to continue off where you left.

**Pull request closed without merging**